src/Polysemy/Internal.hs

```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
+--
+-- @since TODO
+type family InsertedAfter (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedAfter e es (e : r) = e : Append es r
+  InsertedAfter e es (e0 : r) = e0 : InsertedAfter e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' after the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertAfter @e2 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertAfter ::
+  ∀ e es r r' a .
+  r' ~ InsertedAfter e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertAfter = subsume_
+{-# inline insertAfter #-}
+
+------------------------------------------------------------------------------
+-- | Finds the effect @e@ in the stack and inserts the effects @es@ before it.
+--
+-- @since TODO
+type family InsertedBefore (e :: Effect) (es :: EffectRow) (r :: EffectRow) :: EffectRow where
+  InsertedBefore e es (e : r) = Append es (e : r)
+  InsertedBefore e es (e0 : r) = e0 : InsertedBefore e es r
+
+------------------------------------------------------------------------------
+-- | Introduce a set of effects into 'Sem' before the position of the provided
+-- effect @e@. This is intended to be used with a type application:
+--
+-- @
+-- let
+--   sem1 :: Sem [e1, e2, e3, e4, e5] a
+--   sem1 = insertBefore @e5 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertBefore ::
+  ∀ e es r r' a .
+  r' ~ InsertedBefore e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertBefore = subsume_
+{-# inline insertBefore #-}
+
+
+
+data SList l where
+  SEnd  :: SList '[]
+  SCons :: SList xs -> SList (x ': xs)
+
+class KnownList l where
+  singList :: SList l
+
+instance KnownList '[] where
+  singList = SEnd
+  {-# INLINE singList #-}
+
+instance KnownList xs => KnownList (x ': xs) where
+  singList = SCons singList
+  {-# INLINE singList #-}
+
+class Tide (n :: Nat) l where
+  singTide :: SList l
+
+instance {-# OVERLAPPING #-} (l ~ '[]) => Tide 0 l where
+  singTide = SEnd
+
+instance (Tide (n - 1) xs, l ~ (x ': xs)) => Tide n l where
+  singTide = SCons (singTide @(n - 1))
+
+class Build (l :: [a]) (s :: [a]) (r :: [a]) (ix :: Nat) (head :: [a]) (returnType :: [a]) where
+  singBuild :: SList l
+
+instance (l ~ '[]) => Build l r r ix head return where
+  singBuild = SEnd
+
+instance {-# INCOHERENT #-} (Build l1 s r ix head return, l ~ (e ': l1))
+      => Build l (e ': s) r ix head return where
+  singBuild = SCons (singBuild @_ @l1 @s @r @ix @head @return)
+
+instance {-# INCOHERENT #-} TypeError (InsertAtFailure r ix head return)
+       => Build l s r ix head return where
+  singBuild = undefined
+
+type family InsertAtUnprovidedIndex where
+  InsertAtUnprovidedIndex = TypeError ('Text "insertAt: index must be provided")
```

Comment:
`Tide` and `Build` were placeholder names, and should be replaced. My suggestions are`KnownListOfLength` and `KnownPrefix`.

`Tide` isn't `insertAt`-specific, and should reside in `Polysemy.Internal.Membership`. A silly thing here is that the only thing that makes `Build` `insertAt` specific is the need to propagate information for error messages. I suppose that can't be helped.